### DISCUSSION TITLE
Dogecoin Core or Libdogecoin integration on MSVC for app development

### DISCUSSION THREAD

**Discussion by InTimeTechnologies** [dogecoin/dogecoin#3359] [Help needed] [answered]
Hello everyone,

**Issue**
I am using Windows 11, Visual Studio and MSVC to develop an app for my respective OS. I have tried to get a Visual Studio solution up and running with Dogecoin Core and Libdogecoin and failed miserably for more than a year now. I worked around it by running a local node and have my code talk to the node locally. However, this forced my program to lose efficiency and control over some security measures I wanted to implement. Libdogecoin errors trying to find LibEvent despite building from source and providing the path to it and Dogecoin Core building from source does not return .lib files I can use for my program project. Also, the libraries seem to be Linux specific.

**Question**
How can I link to Dogecoin Core or Libdogecoin to my C/C++ Visual Studio solution? I want to have access to the functions it has internally so I can make my app best suited, safer, optimized and user friendly to the public. For example, I want to show my user how much time is left to finish synchronizing the blockchain, but I can't because the commands return errors until it is done synchronizing. I also want to have a standalone server that can return encrypted wallet passphrases and control the node from my code and be notified when the node shuts down and turns on, but I can't. Instead, I have to track these Windows OS processes that can be terminated at will by the user. The experience has not been fun despite committing to the design decision and actually finishing the app.

This is my first time I post on Github under any discussion section. If there is such a thing as an etiquette and broke it, I apologize in advance.

Thank you.

**Comment by patricklodder** [comment, marked as answer]
> How can I link to Dogecoin Core or Libdogecoin to my C/C++ Visual Studio solution?

Dogecoin Core is not a library, so doing any tight coupling by reusing the source code can be risky, because it doesn't export an API cleanly.

Every release does provide script verification routines as `dogecoinconsensus-0.dll`, but to my knowledge there are no production implementations with that at this time - probably because it is not feature-complete to even parse a transaction. It's also not seen any development for a decade.

Instead, look at Dogecoin Core as an isolated service that is capable of reaching a sovereign, validated view of the chain that you can then consume either through the p2p interfaces (for example using `libdogecoin`) or through the RPC server. For either method, this means loose coupling with a `dogecoind`, which is great because it means you can drop-in replace whenever we do a minor release without needing to worry much about rewriting your interfaces. In such a setting, the upside of using the p2p protocol is that you will have complete control versus the upside of using the RPC is that - at least for simple cases - you'll need to implement less blockchain logic.

There is a feature target for `libdogecoin` to have both SPV and full validation implemented, but as long as that is not there and has not been field-hardened, you'll still need a trusted fully validated block source, for example a Dogecoin Core instance that you have full control over, to filter out any potential attacks. As soon as that functionality is released, you can drop the "run a Dogecoin Core instance" requirement and only depend on `libdogecoin`, so implementing that is probably your best bet at this point if you're bound to MSVC.

If your goal is to have such independent binary (eventually) for your front-end application, I think it is wise to ask specific implementation questions to https://github.com/dogecoinfoundation/libdogecoin, so that the people that develop that can help you realize your goals in a safe way.

I'll attempt to answer questions directly related to communicating with the Dogecoin Core process, though remember that this should only be a temporary situation until `libdogecoin` can give you the full functionality you need:

> [..] I can't because the commands return errors until it is done synchronizing

The state of Dogecoin Core is discoverable through RPC (by querying for example `getblockchaininfo`) - as long as you allow some back-off time - as follows:

1. If no reply on the RPC port is given, the process is either down or shutting down. So back off for a longer time, say, a minute, and try to restart the node (this is also where you can poll OS processes if you want to have more control.)
2. If you get error `RPC_IN_WARMUP(-28)` it means the node is starting and you can simply back off for a short time and re-query until it no longer gives you -28
3. If you get a reply on `getblockchaininfo`, the node has started (but does not have to be fully synchronized yet), so now you can make an estimate about remaining time by: 
    - checking whether `.initialblockdownload` is true
    - estimating the total # of blocks at this particular date/time
    - comparing that against `.headers`
    - comparing that against `.blocks`
    - comparing these values over time to get a rough estimate of the time left

> be notified when the node shuts down and turns on

You can use the check for the above described 3 states to detect state in a separate thread.

> This is my first time I post on Github under any discussion section. If there is such a thing as an etiquette and broke it, I apologize in advance.

This is fine, no worries.

**Comment by edtubbs** [comment]
Check out the [getting_started.md](https://github.com/dogecoinfoundation/libdogecoin/blob/0.1.3-dev/doc/getting_started.md) guide if you haven't already. I use Visual Studio Code for libdogecoin development on Windows. Install the necessary dependencies/extensions, such as Visual Studio Build Tools and CMake. Look for "For Windows" in `getting_started.md` for the list.

Once the extensions are installed, open the libdogecoin directory in Visual Studio. It will automatically detect the `CMakeLists.txt` file at the root of the project. There are various CMake targets available, including `tests` and CLI tools, which you can build directly.

Alternatively, you can execute build commands using the Visual Studio Developer Command Prompt; look for "For Windows" in `getting_started.md` for guidance. The project builds libevent during compilation as well as target executables and a statically linked `dogecoin.lib` by default.

As Patrick suggests, let's continue this discussion on [the libdogecoin GitHub repository](https://github.com/dogecoinfoundation/libdogecoin). We'll be glad to assist you there.
